File: JavaScript/0977-有序数组的平方.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
function sortedSquares(nums) {
  const n = nums.length,
    result = new Array(n);
  let l = 0,
    r = n - 1,
    left = 0,
    right = 0;

  for (let i = n - 1; i >= 0; --i) {
    left = nums[l] ** 2;
    right = nums[r] ** 2;

    if (left > right) {
      result[i] = left;
      ++l;
    } else {
      result[i] = right;
      --r;
    }
  }

  return result;
}

File: C/0977-有序数组的平方.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* sortedSquares(int* nums, int numsSize, int* returnSize) {
  *returnSize = numsSize;

  int* result = malloc(sizeof(int) * numsSize);
  int index = numsSize - 1;
  int left = 0, right = numsSize - 1;

  while (left <= right) {
    if (abs(nums[left]) > nums[right]) {
      result[index--] = nums[left] * nums[left];
      left++;
    } else {
      result[index--] = nums[right] * nums[right];
      right--;
    }
  }
  return result;
}
